[needs-docs] Refactoring of all help calls, deprecating QgsContextHelp+Helpviewer

jgrocha opened the pull request:
## Description

This PR drops support for the outdated context help dialogs. The discussion about it has been summarized by @DelazJ in his recent email:
http://osgeo-org.1560.x6.nabble.com/Qgis-community-team-Let-s-bring-documentation-at-the-heart-of-QGIS-Desktop-td5329752.html

The discussion still goes on [QEP](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/51) and [QGIS3_UIX](https://github.com/qgis/qgis3_UIX_discussion/issues/15).

## State of the changes

This PR just eliminates the `QgsContextHelp` in favour of the new `QgsHelp`. Calls to help are (for now) just based on a URL suffix. The support on the server side and the necessary improvements on the desktop side can be done later. This PR just cleans the deprecated `QgsContextHelp`, but it is a necessary step towards the new general help architecture.

## Call pattern (CPP)

I've changed the pattern used to call the help to use the QT signal [`helpRequested`](http://doc.qt.io/qt-5/qdialogbuttonbox.html#helpRequested) with an explicit `connect()`.

Instead of an inline call on the `*.h`:
```cpp
void on_buttonBox_helpRequested() { QgsContextHelp::run( metaObject()->className() ); }
```
I've changed that to:
```cpp
void helpRequest();
```
and changed (`*.cpp`) to:
```cpp
QgsClass::QgsClass(...)
{
  setupUi( this );
  connect( buttonBox, &QDialogButtonBox::helpRequested, this, &QgsClass::helpRequest 
  (...)
);

void QgsClass::helpRequest()
{
  QgsHelp::openHelp( "managing_data_source/opening_data.html#connecting-to-web-services" );
}
```

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

## Tests

For now, there is a simple script called `scripts/chkdoclink.sh` that checks if links provided in the code already exist in the documentation.

The new help infrastructure is not completed. Unit tests should be added later to guarantee that call to help are properly handled when the documentation is either online or offline.

The current CI configuration has problems when open a browser. The problem maybe related with chromium usage by a root user or the lack of a properly display configured for the browser. 

I was not able to solve the problem. Help is much appreciated to solve this issue, but it is not urgent. I've changed `tests/src/python/test_console.py` to make sure the help call is not issued. If a help call is used in the current CI configuration, it will fail. The error is:

```
[41461:41461:0810/001647:FATAL:setuid_sandbox_client.cc(282)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /usr/lib/chromium-browser/chrome-sandbox is owned by root and has mode 4755.
Aborted (core dumped)
Xlib:  extension "RANDR" missing on display ":99".
```

jgrocha committed:
Refactoring of all help calls, deprecating QgsContextHelp and the former helviewer

DelazJ commented on src/gui/ogr/qgsnewogrconnection.cpp:
```diff
@@ -138,4 +139,9 @@ void QgsNewOgrConnection::on_btnConnect_clicked()
 
+void QgsNewOgrConnection::helpRequest()
+{
+  QgsHelp::openHelp( "managing_data_source/opening_data.html#connecting-to-ogr" );
```
This link does not exist. And I don't know to which dialog this button belongs to point you to the right link in doc

jgrocha replied:
@DelazJ I don't think that we have documentation for this. You can reach this dialog when adding a vector layer, then select the database radio button when then New or Edit connection.

![ogr-connection](https://user-images.githubusercontent.com/163681/29139730-35f53424-7d48-11e7-99ef-6d39fcae5c25.png)

DelazJ replied:
Ok. then we can either use:
* http://docs.qgis.org/2.18/en/docs/user_manual/managing_data_source/opening_data.html#loading-a-layer-from-a-file which describes the path to access to that dialog (still in PR2000)
* or http://docs.qgis.org/2.18/en/docs/user_manual/managing_data_source/opening_data.html#creating-a-stored-connection which already describes the parameters in that dialog (might need to slightly add some cross references).

DelazJ commented on src/gui/qgsnewgeopackagelayerdialog.cpp:
```diff
@@ -473,3 +475,7 @@ bool QgsNewGeoPackageLayerDialog::apply()
 
+void QgsNewGeoPackageLayerDialog::helpRequest()
+{
+  QgsHelp::openHelp( "managing_data_source/opening_data.html#loading-a-layer-from-a-file" );
```
The right link is managing_data_source/create_layers.html#creating-a-new-geopackage-layer. It's not yet available in the doc though (pending [PR #2000](https://github.com/qgis/QGIS-Documentation/pull/2000)).

DelazJ commented on src/gui/qgsnewhttpconnection.cpp:
```diff
@@ -330,3 +331,8 @@ void QgsNewHttpConnection::accept()
+
+void QgsNewHttpConnection::helpRequest()
+{
+  QgsHelp::openHelp( "managing_data_source/opening_data.html#connecting-to-web-services" );
```
Actually, this section is just to list in that page full types of data opened by QGIS. It acts like a shortcut so better use the section that treats the subject and avoid one more click to readers. The most relevant page for this dialog might be somewhere in http://docs.qgis.org/testing/en/docs/user_manual/working_with_ogc/index.html

DelazJ commented on src/gui/qgssqlcomposerdialog.cpp:
```diff
@@ -773,3 +776,8 @@ void QgsSQLComposerDialog::setSupportMultipleTables( bool on, const QString &mai
+
+void QgsSQLComposerDialog::helpRequest()
+{
+  QgsHelp::openHelp( "working_with_ogc/ogc_client_support.html#ogc-wfs" );
```
Not sure this is the right link

DelazJ commented on src/app/qgsfieldcalculator.cpp:
```diff
@@ -482,3 +483,8 @@ void QgsFieldCalculator::setPrecisionMinMax()
+
+void QgsFieldCalculator::helpRequest()
+{
+  QgsHelp::openHelp( "working_with_vector/attribute_table.html#field-calculator" );
```
This is indeed the direct link to the field calculator section but I'd keep using working_with_vector/attribute_table.html#editing-attribute-values because:
* it's one level up,
* has an introduction which explains the other ways to edit attribute data (which do not have a help button associated); it could show user that there are better and quicker way to edit data than the field calculator (calculation bar?)
* and field calculator section is the first one so we are not far from the expected section

DelazJ commented on python/console/console.py:
```diff
@@ -59,7 +59,7 @@ def show_console():
     settings = QgsSettings()
     if settings.value('pythonConsole/contextHelpOnFirstLaunch', True, type=bool):
-        QgsContextHelp.run("PythonConsole")
+        QgsHelp.openHelp("../pyqgis_developer_cookbook/intro.html#python-console")
```
There's also http://docs.qgis.org/2.18/en/docs/user_manual/plugins/python_console.html more detailed (which we could improve with link to PyQGIS Doc) and that I'd rather use

DelazJ reviewed (commented):
I made a review from a doc writer perspective and proposed fixes for some doc hyperlink.
FYI, in my summary email you mention above I have a calc doc at https://framacalc.org/4hipbz1q7t in which, as far as i could do, I provided for each dialog the corresponding hyperlink in doc. It might be of help.

What's the logic you'd like to follow for dialogs that are not yet documented (e.g. qgsarcgisservicesourceselect) ?

jgrocha committed:
SIP bindings

jgrocha committed:
Fix links to current documentation

jgrocha committed:
Replace old QSettings with QgsSettings

jgrocha committed:
Prevent test_console.py to open the browser with help file
Fix link to documentation related with OGR connections

jgrocha committed:
helpviewer application source code removed

m-kuhn commented on src/core/CMakeLists.txt:
```diff
@@ -139,8 +139,7 @@ SET(QGIS_CORE_SRCS
   qgsconditionalstyle.cpp
-  qgscontexthelp.cpp
-  qgscoordinatereferencesystem.cpp
+        qgscoordinatereferencesystem.cpp
```
Indentation

m-kuhn commented on src/core/CMakeLists.txt:
```diff
@@ -554,8 +544,7 @@ SET(QGIS_CORE_MOC_HDRS
   qgsbrowsermodel.h
-  qgscontexthelp.h
-  qgscoordinatereferencesystem.h
+        qgscoordinatereferencesystem.h
```
Indentation

m-kuhn commented on src/core/CMakeLists.txt:
```diff
@@ -770,8 +759,7 @@ SET(QGIS_CORE_HDRS
   qgsconnectionpool.h
-  qgscontexthelp.h
-  qgsconditionalstyle.h
+        qgsconditionalstyle.h
```
Indentation

m-kuhn commented on src/gui/ogr/qgsnewogrconnection.cpp:
```diff
@@ -27,12 +26,14 @@
 
 #include <ogr_api.h>
 #include <cpl_error.h>
+#include <qgshelp.h>
```
Use `""` for internal includes

m-kuhn commented on src/gui/ogr/qgsnewogrconnection.h:
```diff
@@ -45,7 +44,7 @@ class GUI_EXPORT QgsNewOgrConnection : public QDialog, private Ui::QgsNewOgrConn
     void accept() override;
     void on_btnConnect_clicked();
-    void on_buttonBox_helpRequested() { QgsContextHelp::run( metaObject()->className() ); }
+    void helpRequest();
```
What's your opinion on renaming this to `showHelp ()`?

Not sure if this actually should be public API. It could as well be private for me. If it stays public, it needs proper doxygen comments.

m-kuhn commented on src/gui/qgsnewhttpconnection.h:
```diff
@@ -39,12 +41,13 @@ class GUI_EXPORT QgsNewHttpConnection : public QDialog, private Ui::QgsNewHttpCo
     void accept() override;
 
+    /**
+     * Open the help in a browser
```
\since QGIS 3.0

m-kuhn commented on src/gui/qgssublayersdialog.h:
```diff
@@ -19,7 +19,7 @@
 #include <QCheckBox>
 #include <ui_qgssublayersdialogbase.h>
-#include "qgscontexthelp.h"
+#include <qgis_sip.h>
```
`""`

m-kuhn replied:
hmmm is this actually required?

m-kuhn commented on src/providers/arcgisrest/qgsarcgisservicesourceselect.cpp:
```diff
@@ -36,6 +35,7 @@
 #include <QFileDialog>
 #include <QRadioButton>
 #include <QImageReader>
+#include <qgshelp.h>
```
`""`

m-kuhn commented on src/providers/wms/qgswmtsdimensions.h:
```diff
@@ -43,7 +42,6 @@ class QgsWmtsDimensions : public QDialog, private Ui::QgsWmtsDimensionsBase
 
     void selectedDimensions( QHash<QString, QString> &dims );
 
-    void on_buttonBox_helpRequested() { QgsContextHelp::run( metaObject()->className() ); }
```
No replacement for this?

jgrocha replied:
I've made a replacement of all `on_buttonBox_helpRequested()` calls to `showHelp()`. Now all help calls to external documentation are consistent.

m-kuhn reviewed (changes requested):
Hey, thanks a lot. I like where this is going. I left some comments, mostly code-wise. One more thing to note is that the strings should be wrapped in `QStringLiteral`

m-kuhn commented:
Concerning the testing I'm not sure where to go. Testing the functionality of QgsHelp itself with an offline package is a nobrainer. I'm hesitating to put calls to the online help into CI, history shows that depending on external infrastructure makes tests flakey. Maybe we can run these tests and let them never fail but upload the missing docs to a list or an issue queue somewhere.

jgrocha committed:
Change all ocurrences on_buttonBox_helpRequested to showHelp
Add help button to 3 dialogs
Fix missing QStringLiteral
Fix CMakeLists.txt indentation
Fix qhelp.h and qgis_sip.h includes with <>

jgrocha commented:
Thanks @m-kuhn for your review.

It took me some time, but I think I've addressed all issues you mention.

In particular, I've made a lot of changes to make all help calls to external documentation more consistent.

I also added some few Help buttons (3x) that were missing in the interface, but with proper documentation prepared and methods ready to handle the event. Those 3 dialogs are the ones with `*.ui` files changed. I not sure way the help buttons were not there. Probably there is some reason behind that. I do not know. If these buttons are simply missing, I've added them. If they should not be there, I'll remove them.

jgrocha committed:
Added documentation comments on new methods

jgrocha committed:
Update sip files

m-kuhn commented:
Thanks @jgrocha , that was a very good move forward!
